### PR TITLE
Improve location handling

### PR DESCRIPTION
**Pull request by nolash**
This PR follows from work migrating location strings in `public.user._location` to actual coordinates. It enables hierarchical relations of locations, aswell as preserving references to external objects (like osm).

A `user_extension_association_table` has been added for this relation, which also could link other user information later on that we would want to extract from the user table in a future cleanup.

### CODE REVIEW

File: app/migrations/versions/241af81119c4_extended_location_information.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+location_external_source_options = ('OSM', 'GEONAMES')
+location_external_source_enum = sa.Enum(*location_external_source_options, name='location_external_source_enum')
+
+ 
+def upgrade():
+
+    # hierarchically nested location resources
+    op.create_table('location',
+            sa.Column('id', sa.Integer(), nullable=False),
+            sa.Column('parent_id', sa.Integer()),
```
Comment (BenSchZA) [inline app/migrations/versions/241af81119c4_extended_location_information.py:28]: Should this be nullable?

File: app/run.py [hunk 2]
```diff
@@ -12,4 +12,5 @@
 
 app = create_app()
 
-app.run(debug=True, threaded=True, host='0.0.0.0', port=9000)
+print(parent_dir)
+#app.run(debug=True, threaded=True, host='0.0.0.0', port=9000)
```
Comment (BenSchZA) [inline app/run.py]: What's the reason for this being removed?

**Review by BenSchZA — COMMENTED** [review]
Added a few small comments.

File: app/run.py (hunk 2 above)
Comment (nolash) [reply app/run.py]: shouldn't thanks.

File: app/migrations/versions/241af81119c4_extended_location_information.py (hunk 1 above)
Comment (nolash) [reply app/migrations/versions/241af81119c4_extended_location_information.py:28]: yes, that would be the biggest geographical unit

**Comment by nolash** [comment]
@BenSchZA I decoupled the osm lookup component from the sqlalchemy model package. would you browse through once more please?

File: share/location/osm.py [hunk 3]
```diff
@@ -49,21 +63,21 @@ def osm_get_place_hierarchy(place_id : int):
...
-        # build and perform osm query
-        response = requests.get('{}{}'.format(url, next_place_id), timeout=QUERY_TIMEOUT)
-        if response.status_code != 200:
-            e = LookupError('failed request to openstreetmap; status code {}'.format(response.status_code))
-            raise(e)
-        response_json = json.loads(response.text)
+       
+        response_json = osm_get_detail(next_place_id)
+        current_place_id = next_place_id
+        next_place_id = response_json['parent_place_id']
```
Comment (mango-habanero) [inline share/location/osm.py:97]: Considering the `osm_get_detail` method returns a dict if I am not wrong. I think it is safer to use the `.get()` method which should default to `None` in the event that the key you're looking for is absent. Calling this:
```
response_json['parent_place_id']
```
will raise a `keyError` if its absent.
On one hand I imagine its reasonably safe to assume that the value will always be returned in the response but I prefer the `get` just to be safe.

File: share/location/osm.py [hunk 4]
```diff
@@ -147,3 +160,60 @@ def osm_resolve_name(name, country=DEFAULT_COUNTRY_CODE):
...
     return locations[0]
+
+
+def osm_resolve_coordinates(latitude, longitude):
+  
+    q = {
+        'format': 'json',
+        'lat': latitude,
+        'lon': longitude,
+    }
```
Comment (mango-habanero) [inline share/location/osm.py]: Purely a matter a preference here but using a value such as `q` kind of robs the variable of its contextual meaning, from face value its a bit difficult to tell what `q` is. Seems like its a query parameter so I'd call it that

**Review by mango-habanero — COMMENTED** [review]
Its looks okay to me save for the two comments which are largely matters of preference. Generally curious though, why did you move this to a share dir ?

File: share/location/osm.py (hunk 3 above)
Comment (nolash) [reply share/location/osm.py:97]: I would actually prefer an exception here, as the data coming back should _always_ have this field. If it doesn't, we will catch either wrong assumptions and/or changes in the source data structure.

File: share/location/osm.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+        query['email'] = config.EXT_OSM_EMAIL
+    query_string = urllib.parse.urlencode(query)
+
+    # perform osm query
+    url = 'https://nominatim.openstreetmap.org/search?' + query_string
+    try:
+        response = requests.get(url, timeout=QUERY_TIMEOUT)
+    except requests.exceptions.Timeout:
+        logg.warning('request timeout to openstreetmap; {}:{}'.format(country, name))
+        # TODO: re-insert task
```
Comment (BenSchZA) [inline share/location/osm.py]: What is the rogue TODO doing here?

**Review by BenSchZA — DISMISSED** [review]

**Pull request merged**